include/emp/prefab/ValueBox.hpp

```diff
@@ -0,0 +1,182 @@
+      const std::string & desc,
+      const std::string & inital_value,
+      web::Input input,
+      const std::string & id=""
+    ) : ValueBox(label, desc, id), mainCtrl(input) {
+      view << mainCtrl;
+      mainCtrl.Value(inital_value);
+    }
+  };
+
+  class TextValueControl : public ValueControl {
+    public:
+    TextValueControl(
+      const std::string & label,
+      const std::string & desc,
+      const std::string & value,
+      const std::function<void(const std::string & val)> & onChange = [](std::string val) { ; },
+      const std::string & id=""
+    ) : ValueControl(label, desc, value, web::Input(onChange, "text", ""), id) {
+      mainCtrl.AddAttr("class", "form-control");
+    }
+  };
+
+  class BoolValueControl : public ValueBox {
+    public:
+    BoolValueControl(
+      const std::string & label,
+      const std::string & desc,
+      const bool is_checked,
+      const std::function<void(const std::string & val)> & onChange = [](const std::string & val) { ; },
+      const std::string & id=""
+    ) : ValueBox(label, desc, id) {
+      prefab::ToggleSwitch toggle(onChange, "", is_checked);
+      view << toggle;
+    }
+  };
+
+  class NumericValueControl : public ValueControl {
+
+    // Determine the default range by finding the next highest order of magnitude (base 10)
+    inline static std::function<void(
+      const std::string &,
+      const std::string &,
+      emp::web::Input &
+    )> applyDefaultRange = [](
+      const std::string & value,
+      const std::string & type,
+      emp::web::Input & in
+    ) {
+      if(type == "float" || type == "double") {
+        double val = emp::from_string<double>(value);
+        int max = (abs(val) <= 1) ? 1 : (int)pow(10, ceil(log10(abs(val))));
+        int min = (val >= 0) ? 0 : -max;
+        double step = max/100.0;
+        in.Max(max);
+        in.Min(min);
+        in.Step(step);
+      } else {
+        int val = emp::from_string<int>(value);
+        int max = (abs(val) <= 10) ? 10 : (int)pow(10, ceil(log10(abs(val))));
+        int min = (val >= 0) ? 0 : -max;
+        int step = (int)fmax(max/100.0, 1);
+        in.Max(max);
+        in.Min(min);
+        in.Step(step);
+      }
+    };
+
+    public:
+    NumericValueControl(
+      const std::string & label,
+      const std::string & desc,
+      const std::string & value,
+      const std::string & type,
+      const std::function<void(const std::string & val)> & onChange = [](const std::string & val) { ; },
+      const std::string & id=""
+    ) : ValueControl(label, desc, value, web::Input([](const std::string & val){ ; }, "range", ""), id) {
+      mainCtrl.AddAttr("class", "form-range");
+      web::Input temp(mainCtrl);
+      web::Input number_box([slider=temp, onChange](const std::string & val) mutable {
+        // Lambdas must be marked mutable since .Value is not a const function
+        // Warning: referenced components/functions must be captured by value at
```

Comment:
warning makes me think something's still broken? I think the way you did it is fixed so might relabel slightly

Reply:
Maybe I should just change it to a note. The pattern in most lambdas seemed to be capturing by reference but because  the function `onChange` and `temp` go out of scope eventually, we need copies. Really it seems like it should be `Input`'s job to prevent dangling references but not sure if that's possible.